datafusion/functions/src/datetime/to_local_time.rs

```diff
@@ -0,0 +1,601 @@
+                Ok(time_value.clone())
+            }
+            // if has timezone, adjust the underlying time value. the current time value
+            // is stored as i64 in UTC, even though the timezone may not be in UTC, so
+            // we need to adjust the time value to the local time. see [`adjust_to_local_time`]
+            // for more details.
+            //
+            // Then remove the timezone in return type, i.e. return None
+            DataType::Timestamp(_, Some(_)) => match time_value {
+                ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(
+                    Some(ts),
+                    Some(tz),
+                )) => {
+                    let adjusted_ts =
+                        adjust_to_local_time::<TimestampNanosecondType>(ts, &tz);
+                    Ok(ColumnarValue::Scalar(ScalarValue::TimestampNanosecond(
+                        Some(adjusted_ts),
+                        None,
+                    )))
+                }
+                ColumnarValue::Scalar(ScalarValue::TimestampMicrosecond(
+                    Some(ts),
+                    Some(tz),
+                )) => {
+                    let adjusted_ts =
+                        adjust_to_local_time::<TimestampMicrosecondType>(ts, &tz);
+                    Ok(ColumnarValue::Scalar(ScalarValue::TimestampMicrosecond(
+                        Some(adjusted_ts),
+                        None,
+                    )))
+                }
+                ColumnarValue::Scalar(ScalarValue::TimestampMillisecond(
+                    Some(ts),
+                    Some(tz),
+                )) => {
+                    let adjusted_ts =
+                        adjust_to_local_time::<TimestampMillisecondType>(ts, &tz);
+                    Ok(ColumnarValue::Scalar(ScalarValue::TimestampMillisecond(
+                        Some(adjusted_ts),
+                        None,
+                    )))
+                }
+                ColumnarValue::Scalar(ScalarValue::TimestampSecond(
+                    Some(ts),
+                    Some(tz),
+                )) => {
+                    let adjusted_ts =
+                        adjust_to_local_time::<TimestampSecondType>(ts, &tz);
+                    Ok(ColumnarValue::Scalar(ScalarValue::TimestampSecond(
+                        Some(adjusted_ts),
+                        None,
+                    )))
+                }
+                ColumnarValue::Array(array) => {
+                    fn transform_array<T>(
+                        array: &ArrayRef,
+                        tz: &str,
+                    ) -> Result<ColumnarValue>
+                    where
+                        T: ArrowTimestampType,
+                    {
+                        let array = as_primitive_array::<T>(array)?;
+                        let array: PrimitiveArray<T> =
+                            array.unary(|ts| adjust_to_local_time::<T>(ts, tz));
```

Comment:
minor remark: You can have both `array` definitions in the same definition to avoid shadowing
```suggestion
                        let array = as_primitive_array::<T>(array)?.unary(|ts| adjust_to_local_time::<T>(ts, tz));
```
(my linting might be off :p) 

Thinking outloud here, but perhaps it's better to follow the same naming convention as for the scalar types. i.e:
```suggestion
                        let array = as_primitive_array::<T>(array)?;
                        let adjusted_array = .unary(|ts| adjust_to_local_time::<T>(ts, tz));
```

Reply:
Thank you for the suggestion! I added error handling in https://github.com/apache/datafusion/pull/11347/commits/afe619f057eb4c7830f1fa01933527d21fa30d61. As a result, this part of the code got rewritten.